doc/developer/design/20230209_invalid_accumulations.md

```diff
@@ -0,0 +1,78 @@
+
+At a high level, our solution approach is to add a post-processing operator immediately after each reduction operator that evaluates if an error occurred during reduction and then produces an error in the appropriate error stream if so. This design avoids changing reduction operators to potentially create error streams, thus reducing the complexity of the change as well as avoiding compromises with raw reduction performance. Two important aspects need to be tackled for this solution approach to work: (1) We need to find a method to concretely implement the post-processing operator; (2) We need to devise a strategy to externalize the information that an error occurred in the output of the reduction operator.
+
+Regarding point (1), the post-processing operator could be implemented as a special-purpose operator in rendering or as an MFP expression with the relevant error processing code encoded as an `MirScalarExpr`. The former option maintains all the implementation complexity within rendering. An argument can be made against this choice, since the error conditions are based on SQL-level semantics and it would be meaningful to encode them at a higher level of abstraction. The latter option embraces error handling as a SQL-level construct, but may suffer from lower evaluation performance than a special-purpose operator introduced during rendering.
+
+Regarding point (2), we note that: (a) The error detection strategy for invalid accumulations is specific to the reduction type; and (b) the necessary information for error detection, e.g., row multiplicities, is available only at the input to a reduce closure. Because of (a), ideally each reduction operator should output, along with an aggregate, an additional field that indicates its validity, i.e., we represent whether an error was detected as an additional field per aggregate row. Due to (b), we need to represent this additional output also in arrangements, complicating the use of a traditional `Result` type. Thus, we propose to employ an extra `Row` column that contains either `Datum::True` or `Datum::False` indicating whether the aggregate is valid or invalid, respectively. The post-processing operator can then produce errors upon seeing `Datum::False` in this validity column or, otherwise, project out the column.
```

Comment:
Have you considered using an `Datum::String` instead, to encode an error message, rather than just encoding "some error occurred"? I'm not sure whether or not this would be necessary, but I can imagine that the post-processing operator lacks information to distinguish between different error conditions.

Reply:
I tried to keep this minimal here because we know the key for which the invalid aggregate was generated and also that the error is in the context of a given dataflow. Given that the column would be used exclusively to flag invalid accumulations, we'd know also the type of error. So there seems to be enough information to construct a meaningful error message in the post-processing step.